### PR TITLE
polaris: init at 0.13.5

### PR DESCRIPTION
**Pull request by pbsds**
###### Description of changes

This PR adds `polaris`, its sub-dependency `polaris-web`, and a nixos module to run it.

image

Polaris is a self-hosted music streaming server, with a web and android frontend. The web frontend is included in the official releases.
https://github.com/agersant/polaris

`polaris-web` lies in a separate github repository, and inspired by the packages `jellyfin` and `jellyfin-web` I chose to build `polaris-web` as a separate package.
`polaris-web` uses an odd versioning schema upstream: "build-X". I chose to reuse these tags. Official `polaris` releases (Github actions CI) automatically fetches and bundles the latest available version of `polaris-web`, however `polaris` has seen no new release since april 2021, while `polaris-web` has since had 5 version bumps. Currently I package the version of 'polaris-web' which is being bundled with the current `polaris` release on Github. Once I have properly tested the newer versions I might upgrade.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by pbsds** [comment]
Result of `nixpkgs-review pr 178052` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>polaris</li>
    <li>polaris-web</li>
  </ul>
</details>

**Review by D4ndellion — CHANGES_REQUESTED** [review]
Generally good (especially for a ~first time contributor), just a few comments/improvements.

If I were you I'd not bother packaging the webclient from source though and just go with the one provided in the release. But it's really up to you

File: nixos/modules/services/misc/polaris.nix [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "polaris";
+        description = "Group under which Polaris is run.";
+      };
+
+      port = mkOption {
+        type = types.int;
```
Comment (legendofmiracles) [inline nixos/modules/services/misc/polaris.nix:28]:
```suggestion
        type = types.port;
```

File: pkgs/servers/polaris/default.nix [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  #  })
+  #];
+
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+  };
+
+  postPatch = ''
+    cp ${./Cargo.lock} Cargo.lock
+  '';
```
[the remark this answers is not in the document]
Comment (happysalada) [reply pkgs/servers/polaris/default.nix:34]: the original repo includes a lockfile also, did the original lockfile give you any trouble ?

File: pkgs/servers/polaris/default.nix [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  # compile-time environment variables
+  POLARIS_WEB_DIR = "${polaris-web}/share/polaris-web";
+  POLARIS_SWAGGER_DIR = "${placeholder "out"}/swagger";
+
+  postInstall = ''
+    cp -r docs/swagger $out/
+  '';
```
[the remark this answers is not in the document]
Comment (pbsds) [reply pkgs/servers/polaris/default.nix:42]: This is a prebuilt swagger index.html++ bundle made available at `http://.../swagger` for viewing the Polaris OpenAPI specification.

[the remark this answers is not in the document]
Comment (dali99) [reply pkgs/servers/polaris/default.nix:42]: I see, they should still go to somewhere else than in the raw out

**Comment by pbsds** [comment]
Sorry for the delay, I've been having issues with the module 'actix-rt' raising a "Too many open files" error during `cargo test`.
This only seems to happen when I run `nix-build` as root on my server.
Running it with my normal user made it pass */shrug*.
*(Even though nix builds as `nixbld1`, it seems to inherit the ulimits of the initiating user?)*

I decided to keep the separate webui package simply because there is a precedent for polaris-web receiving minor bugfixes without there being any version bumps in polaris.
This approach allows us to decouple their versions.

I commit now "fixup" commits for the sake of review, to be squashed when rebasing.

I also now added a nixos test for the nixos module.

File: nixos/modules/services/misc/polaris.nix [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        RestrictAddressFamilies = [ "AF_INET" "AF_INET6" "AF_UNIX" ];
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+
+        SystemCallArchitectures = "native";
+        SystemCallErrorNumber = "EPERM";
+        SystemCallFilter = [
+          "@system-service"
+          "~@cpu-emulation" "~@debug" "~@keyring" "~@memlock" "~@obsolete" "~@privileged" "~@setuid"
+        ];
```
[the remark this answers is not in the document]
Comment (pbsds) [reply nixos/modules/services/misc/polaris.nix]: I mostly lifted these from the jellyfin module. I did comment the ones implicitly enabled when in switched to DynamicUser.

File: pkgs/servers/polaris/default.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (pbsds) [reply pkgs/servers/polaris/default.nix]: The Cargo.lock is now patched during fetchFromGithub using postFetch, removing the need to vendor it.
I left a comment detailing its need.

File: pkgs/servers/polaris/default.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (pbsds) [reply pkgs/servers/polaris/default.nix]: I moved it to /share/polaris-swagger

File: pkgs/servers/polaris/default.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (pbsds) [reply pkgs/servers/polaris/default.nix]: I did also learn about`cargoPatches`, but I decided not to use it as the instability of lock files will likely require us to update the patch each version bump.

**Comment by happysalada** [comment]
This looks good to me.
Would you mind squashing these into 3 commits ?
(one fore the maintainer, one for the package, and one for the module and the test)
From the github UI I can only squash all into 1 commit.

**Review by D4ndellion — APPROVED** [review]

**Comment by happysalada** [comment]
Thanks again!

**Pull request merged**